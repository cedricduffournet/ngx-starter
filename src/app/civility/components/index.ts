export * from './civility-add/civility-add.component';
export * from './civility-update/civility-update.component';
export * from './civility-delete/civility-delete.component';
export * from './civility-form/civility-form.component';
export * from './civility-item/civility-item.component';
export * from './civility-items/civility-items.component';
