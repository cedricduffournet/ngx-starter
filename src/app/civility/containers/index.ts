export * from './civility-add-modal/civility-add-modal.component';
export * from './civility-update-modal/civility-update-modal.component';
export * from './civility-delete-modal/civility-delete-modal.component';
export * from './civility-list-view/civility-list-view.component';
