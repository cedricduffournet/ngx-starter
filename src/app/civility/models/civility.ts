import { schema } from 'normalizr';

export interface Civility {
  id: number;
  name: string;
  code: string;
}

export const civilitySchema = new schema.Entity('civilities');
