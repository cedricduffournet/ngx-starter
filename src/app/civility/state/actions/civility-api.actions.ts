import { props, createAction } from '@ngrx/store';

import { NormalizedData } from '@app/shared/models/normalized.model';

export const loadCivilitySuccess = createAction(
  '[Civility API] Load civility SUCCESS',
  props<{ civilities: NormalizedData }>()
);

export const loadCivilityFailure = createAction(
  '[Civility API] Load civility FAILURE',
  props<{ error: any }>()
);

export const addCivilitySuccess = createAction(
  '[Civility API] Add civility SUCCESS',
  props<{ civility: NormalizedData }>()
);

export const addCivilityFailure = createAction(
  '[Civility API] Add civility FAILURE',
  props<{ error: any }>()
);

export const updateCivilitySuccess = createAction(
  '[Civility/entities] Update civility SUCCESS',
  props<{ civility: NormalizedData }>()
);

export const updateCivilityFailure = createAction(
  '[Civility/entities] Update civility FAILURE',
  props<{ error: any }>()
);

export const deleteCivilitySuccess = createAction(
  '[Civility API] Remove civility SUCCESS',
  props<{ id: number }>()
);

export const deleteCivilityFailure = createAction(
  '[Civility API] Remove civility FAILURE',
  props<{ error: any }>()
);
