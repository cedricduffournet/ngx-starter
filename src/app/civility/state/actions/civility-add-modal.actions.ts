import { props, createAction } from '@ngrx/store';

import { Civility } from '@app/civility/models/civility';

export const addCivility = createAction(
  '[Add civility modal] Add civility',
  props<{ civility: Civility }>()
);
