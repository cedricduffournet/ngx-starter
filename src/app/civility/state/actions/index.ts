import * as CivilityUpdateModalActions from './civility-update-modal.actions';
import * as CivilityAddModalActions from './civility-add-modal.actions';
import * as CivilityDeleteModalActions from './civility-delete-modal.actions';
import * as CivilityListViewActions from './civility-list-view.actions';
import * as CivilityApiActions from './civility-api.actions';

export {
  CivilityListViewActions,
  CivilityApiActions,
  CivilityUpdateModalActions,
  CivilityAddModalActions,
  CivilityDeleteModalActions
};
