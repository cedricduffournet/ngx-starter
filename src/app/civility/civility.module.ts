import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { CivilityEffects } from '@app/civility/state/effects';
import {
  CivilityAddComponent,
  CivilityUpdateComponent,
  CivilityDeleteComponent,
  CivilityFormComponent,
  CivilityItemComponent,
  CivilityItemsComponent
} from '@app/civility/components';
import {
  CivilityAddModalComponent,
  CivilityUpdateModalComponent,
  CivilityDeleteModalComponent,
  CivilityListViewComponent
} from '@app/civility/containers';
import { reducers } from '@app/civility/state/reducers';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    EffectsModule.forFeature([CivilityEffects]),
    StoreModule.forFeature('civilities', reducers)
  ],
  declarations: [
    CivilityListViewComponent,
    CivilityAddComponent,
    CivilityUpdateComponent,
    CivilityDeleteComponent,
    CivilityFormComponent,
    CivilityItemComponent,
    CivilityItemsComponent,
    CivilityAddModalComponent,
    CivilityUpdateModalComponent,
    CivilityDeleteModalComponent
  ],
  entryComponents: [
    CivilityUpdateModalComponent,
    CivilityAddModalComponent,
    CivilityDeleteModalComponent
  ],
  exports: [CivilityListViewComponent]
})
export class CivilityModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
