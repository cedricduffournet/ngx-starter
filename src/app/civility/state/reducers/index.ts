import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers
} from '@ngrx/store';

import * as fromRoot from '@app/core/state/reducers';
import * as fromCivilityCollection from '@app/civility/state/reducers/civility-collection.reducer';
import * as fromCivilityEntities from '@app/civility/state/reducers/civility-entities.reducer';
import * as fromAuth from '@app/authentication/state/reducers';
import { Civility } from '@app/civility/models/civility';
import { Authorization } from '@app/core/models/authorization.model';

export interface CivilitiesState {
  collection: fromCivilityCollection.State;
  civilities: fromCivilityEntities.State;
}

export interface State extends fromRoot.State {
  civilities: CivilitiesState;
}

export function reducers(state: CivilitiesState | undefined, action: Action) {
  return combineReducers({
    collection: fromCivilityCollection.reducer,
    civilities: fromCivilityEntities.reducer
  })(state, action);
}

export const selectCivilitiesState = createFeatureSelector<
  State,
  CivilitiesState
>('civilities');

export const getCivilityEntitiesState = createSelector(
  selectCivilitiesState,
  state => state.civilities
);

export const getCivilityEntities = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getEntities
);

export const getCivilityEntitiesUpdating = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getUpdating
);

export const getCivilityEntitiesUpdated = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getUpdated
);

export const getCivilityCollectionState = createSelector(
  selectCivilitiesState,
  state => state.collection
);
export const getCivilityIds = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getIds
);
export const getCivilityCollectionAdding = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getAdding
);
export const getCivilityCollectionAdded = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getAdded
);
export const getCivilityCollectionDeleting = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getDeleting
);
export const getCivilityCollectionDeleted = createSelector(
  getCivilityCollectionState,
  fromCivilityCollection.getDeleted
);

export const getCivilities = createSelector(
  getCivilityEntities,
  getCivilityIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const getSelectedCivilityId = createSelector(
  getCivilityEntitiesState,
  fromCivilityEntities.getSelectedId
);

export const getSelectedCivility = createSelector(
  getCivilityEntities,
  getSelectedCivilityId,
  (civilityEntities, civilityId): Civility => {
    return civilityEntities[civilityId];
  }
);

export const canUpdateCivility = createSelector(
  fromAuth.getAuthorized(['ROLE_CIVILITY_EDIT']),
  canUpdate => canUpdate
);

export const canDeleteCivility = createSelector(
  fromAuth.getAuthorized(['ROLE_CIVILITY_DELETE']),
  canDelete => canDelete
);

export const canCreateCivility = createSelector(
  fromAuth.getAuthorized(['ROLE_CIVILITY_CREATE']),
  canCreate => canCreate
);

export const getCivilityAuthorization = createSelector(
  canUpdateCivility,
  canDeleteCivility,
  canCreateCivility,
  (canUpdate, canDelete, canAdd): Authorization => {
    return {
      update: canUpdate,
      delete: canDelete,
      create: canAdd
    };
  }
);
