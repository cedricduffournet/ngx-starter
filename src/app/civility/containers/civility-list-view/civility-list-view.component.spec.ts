import { TestBed, ComponentFixture } from '@angular/core/testing';

import { Store } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { TranslateModule } from '@ngx-translate/core';

import {
  CivilityItemComponent,
  CivilityItemsComponent
} from '@app/civility/components';
import { CivilityListViewComponent } from '@app/civility/containers';
import * as fromCivilities from '@app/civility/state/reducers';
import { CivilityListViewActions } from '@app/civility/state/actions';
import { SharedModule } from '@app/shared/shared.module';

describe('CivilityListViewComponent', () => {
  let fixture: ComponentFixture<CivilityListViewComponent>;
  let component: CivilityListViewComponent;
  let store: MockStore<fromCivilities.State>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        CivilityListViewComponent,
        CivilityItemsComponent,
        CivilityItemComponent
      ],
      imports: [SharedModule, TranslateModule.forRoot()],
      providers: [
        provideMockStore({
          selectors: [
            {
              selector: fromCivilities.getCivilities,
              value: []
            },
            {
              selector: fromCivilities.getCivilityAuthorization,
              value: {
                create: true,
                delete: true,
                update: true
              }
            }
          ]
        })
      ]
    });

    fixture = TestBed.createComponent(CivilityListViewComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    spyOn(store, 'dispatch');
  });

  it('should be created', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch loadCivilities on init', () => {
    const action = CivilityListViewActions.loadCivilities();
    fixture.detectChanges();
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch showAddCivilityModal on add event', () => {
    const action = CivilityListViewActions.showAddCivilityModal();
    component.onAdd();
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch showUpdateCivilityModal on update event', () => {
    const action = CivilityListViewActions.showUpdateCivilityModal();
    component.onUpdate(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch selectCivility on update event', () => {
    const action = CivilityListViewActions.selectCivility({ id: 1 });
    component.onUpdate(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch showDeleteCivilityModal on update event', () => {
    const action = CivilityListViewActions.showDeleteCivilityModal();
    component.onDelete(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch selectCivility on deletee event', () => {
    const action = CivilityListViewActions.selectCivility({ id: 1 });
    component.onDelete(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
