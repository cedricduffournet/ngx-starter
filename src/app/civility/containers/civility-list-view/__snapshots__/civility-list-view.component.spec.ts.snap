// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CivilityListViewComponent should be created 1`] = `
<app-civility-list-view
  authorization$={[Function Store]}
  civilities$={[Function Store]}
  loggedUser$={[Function Store]}
  store={[Function MockStore]}
>
  <app-page-inner>
    <div
      class="px-3 py-3 "
    >
      <div
        class="page-section"
      >
        <app-civility-items
          ng-reflect-authorization="[object Object]"
          ng-reflect-civilities=""
        >
          <section
            class="bg-white border-left border-right border-bottom rounded"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table table-hover"
              >
                <thead>
                  <tr>
                    <th
                      class="align-middle"
                      ng-reflect-translate=""
                      translate=""
                    >
                       ABBREVIATION 
                    </th>
                    <th
                      class="align-middle"
                      ng-reflect-translate=""
                      translate=""
                    >
                       NAME 
                    </th>
                    <th
                      class="align-middle text-right"
                    >
                       Actions 
                    </th>
                  </tr>
                </thead>
                <tbody>
                  
                </tbody>
              </table>
            </div>
          </section>
          
          <app-button-circle-right
            label="+"
            ng-reflect-label="+"
          >
            <div>
              <button
                class="btn-circle btn-xl btn btn-absolute-corner-right btn-success"
                ng-reflect-klass="btn-circle btn-xl btn btn-abso"
                ng-reflect-ng-class="btn-success"
                type="button"
              >
                 + 
              </button>
            </div>
          </app-button-circle-right>
        </app-civility-items>
      </div>
    </div>
  </app-page-inner>
</app-civility-list-view>
`;
